### ISSUE TITLE
[question] Creating a copy of some package with different version (and maybe name)

### ISSUE THREAD

**Issue by lo1ol** [conan-io/conan#15955]
### What is your question?

Hi!

In first version of conan we had `alias` command, which allows us to create aliases to packages, but in conan 2.0 we don't.

Right now, we are thinking how to promote our release candidate packages to release packages placed in another repo. Can you please hint us some of the best practices of creating copies of similar packages, that differ only in version or name? Binaries inside packages should not be changed.

### Have you read the CONTRIBUTING guide?

- [ ] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @lo1ol 

Conan 2 has been designed following well known devops practices, like package immutability, and dropping previously bad practices. Creating packages with different versions and names from the same binaries is an anti-pattern, and shouldn't be done.  One given package with the same binaries should have one name and version (and revision, that could be considered part of the version).

A different aspect is that sometimes it is desired to control which packages are part of a dependency graph. This can be controlled in different ways:
- Providing lockfiles to control the versions and revisions of dependencies and transitive dependencies
- Using different server repositories to store packages at different "stages". If you  have some packages that you want to test for the next release, you put those packages in a server repo named "release_candidate" or the like, then configure the remote, use those packages and test them. When things work, the package must not be renamed, but instead "promoted", which is basically copied to another "production" or "release" server repo, under the exact same name and version.
- Different versioning schemes can also be followed, in combination with the above, for example Conan 2 implement "prereleases" which is the standard versioning mechanism for release candidates, and then the ``core.version_ranges:resolve_prereleases`` configuration can be used to opt-in to use those prereleases in a build, without the need to store them in yet another server repo.

**Comment by lo1ol** [comment]
I got you. But I still have some questions about real implementation. Would you please describe it with maybe some examples?

In first cases I have struggles with naming if take your approaches. For example. Here it out approach:
1. we build some product and regularly push package with it to our testing repo. Naming of this package is `package/0.0.1-commit-hash`
2. After some time we decided to make a release candidate of some package and call it. So, we change the version file in our project and build release package, that called `package/1.0.0-rc1` or just  `package/1.0.0` and put it in release condidate repo. Also we still wanted to keep commit hash in version, like we used-to because it's convenient for us (for example `package/1.0.0-rc1-abcdef`). Is it a bad practice? Is there another approaches where to keep this info?
3. accidentally we found out that there is some bugs in it. So, we fix it and decided to create another rc package. How it should be called? `package/1.0.0-rc2-commit-hash`? Notice, that testing packages between this release condidates will be called `package/1.0.0-commit-hash` (not `package/0.0.1-commit-hash`). It's kind of weird
4.  this package pass everything. So, we should promote last rc package to release package, that called `package/1.0.0-rc2-commit-hash`, without changing its name?

Btw, about lock-files. I have  some misunderstanding with them. We build our product with a lot of profiles and each product could be building with different settings. How lock files could help me, because I have to get lock-files for all combinations of my profiles and settings and keep it somewhere? Of course I could write some script, that did this for me, but are you sure that it's normal practice? Or I don't got the idea of using lock files?

**Comment by memsharded** [comment]
> After some time we decided to make a release candidate of some package and call it. So, we change the version file in our project and build release package, that called package/1.0.0-rc1 or just package/1.0.0 and put it in release condidate repo. Also we still wanted to keep commit hash in version, like we used-to because it's convenient for us (for example package/1.0.0-rc1-abcdef). Is it a bad practice? Is there another approaches where to keep this info?

Depending on the usage of the commit hash, it might be possible:
- To store it in the recipe revision with ``revision_mode = "scm"`` attribute in recipes
- To store it explicitly in the ``conandata.yml``. This can be done in the ``export()`` method, or with a hook to apply to all recipes.

The problem with having commit hashes in versions is that they are not ordered, so it is a bad choice for version resolution. It could be added to "build-metadata", like ``pkg/1.0.0-rc1+sha``, but it should imply that a new sha commit gets a new ``rc2`` pre-release version too, otherwise a version range will resolve to the latest **alphabetical** sha, which will be incorrect.

> accidentally we found out that there is some bugs in it. So, we fix it and decided to create another rc package. How it should be called? package/1.0.0-rc2-commit-hash? Notice, that testing packages between this release condidates will be called package/1.0.0-commit-hash (not package/0.0.1-commit-hash). It's kind of weird

There are 3 branches of development now then? Not very clear. If you keep rolling the dev ``0.0.1`` and you haven't finalized the ``rc``, then it should be impossible to have a ``1.0.0-sha`` yet? 

> this package pass everything. So, we should promote last rc package to release package, that called package/1.0.0-rc2-commit-hash, without changing its name?

The release candidate is yet a pre-version, not the final one. Once the decision to move to ``1.0.0`` is done, then the final release ``pkg/1.0.0`` should be created, uploaded, tested, and promoted in the same way the release candidates were. There are many situations in which there are bug fixes between the latest release candidate and the final release, without another release candidate. The reason it is very important to do this is because there can be version-associated bugs, like recipe logic with: ``if self.version <= xxx: => do something``, and this logic can change, be buggy in presence of pre-releases. Version checking logic in recipes is very, very common. The only way to guarantee that the ``1.0.0`` package works correctly is to fully build and test it with the final real value of the version, not convert a previous binary with a different version. This is a lesson learned while Conan 1, as this is also a cause of very challenging things to debug and understand, typically affecting at runtime/production.


> Btw, about lock-files. I have some misunderstanding with them. We build our product with a lot of profiles and each product could be building with different settings. How lock files could help me, because I have to get lock-files for all combinations of my profiles and settings and keep it somewhere? Of course I could write some script, that did this for me, but are you sure that it's normal practice? Or I don't got the idea of using lock files?

Lockfiles in Conan 2 can contain/lock dependencies for different configurations/profiles in the same single lockfile. If you want to keep consistency between dependencies for different platforms, definitely capturing the lockfile for all the configurations and using it is recommended. Consider that your product depends on ``zlib/[>1.2 <2]``, and you need to build it for different platforms. If you don't capture a lockfile, it might be possible that your Windows binary will be built with zlib/1.2.13, and your Linux one with zlib/1.3, just because it happened that the ``zlib/1.3`` package was uploaded to your server in the middle of the 2 builds. If you plan to build your product for different configs, iterating those profiles first with ``conan lock create`` would be recommended.

**Comment by lo1ol** [comment]
> There are 3 branches of development now then?

Yeah. We are planning to have 3 repo: testing, rc, release.

> If you keep rolling the dev 0.0.1 and you haven't finalized the rc, then it should be impossible to have a 1.0.0-sha yet?

yes

Got you idea about lock files. We check it:) Thank you!

**Comment by memsharded** [comment]
> Got you idea about lock files. We check it:) Thank you!

Happy to help 🙂 thanks for your feedback!

Maybe this ticket can be closed now as responded? You can re-open or create a new ticket at any time for any further question.
